# Konten
1. Overview Proyek
2. Objektif Proyek
3. Deskripsi Data
4. Import library & Eksplorasi Data
5. Pengolahan Data
6. Kesimpulan 

# Overview Proyek
󠀰Virus corona yang kehadirannya mengejutkan seluruh dunia telah mengubah rutinitas harian semua orang.󠀲󠀡󠀠󠀦󠀥󠀨󠀧󠀢󠀳󠀰 Kini, warga kota tidak lagi menghabiskan waktu luangnya di luar rumah seperti pergi ke kafe dan mal. Mereka lebih sering berada di rumah, menghabiskan waktunya dengan membaca buku.󠀲󠀡󠀠󠀦󠀥󠀨󠀧󠀣󠀳󠀰 Hal ini pun mendorong perusahaan startup untuk mengembangkan aplikasi baru bagi para pecinta buku.󠀲󠀡󠀠󠀦󠀥󠀨󠀧󠀤󠀳

󠀰Kamu telah diberi sebuah database dari salah satu perusahaan yang bersaing dalam industri ini.󠀲󠀡󠀠󠀦󠀥󠀨󠀧󠀥󠀳󠀰 Database tersebut berisi data tentang buku, penerbit, penulis, serta rating dan ulasan pelanggan atas buku terkait.󠀲󠀡󠀠󠀦󠀥󠀨󠀧󠀦󠀳󠀰 Informasi ini akan digunakan dalam membuat penawaran harga untuk sebuah produk baru.

# Objektif Proyek
1. Hitung jumlah buku yang dirilis setelah tanggal 1 Januari 2000.󠀲󠀡󠀠󠀦󠀥󠀩󠀠󠀩󠀳
2. Hitung jumlah ulasan pengguna dan rating rata-rata untuk setiap buku.󠀲󠀡󠀠󠀦󠀥󠀩󠀡󠀠󠀳
3. Identifikasi penerbit yang telah menerbitkan jumlah buku terbanyak dengan jumlah halaman lebih dari 50 (hal ini akan membantumu dalam mengecualikan brosur dan publikasi serupa dari analisismu).󠀲󠀡󠀠󠀦󠀥󠀩󠀡󠀡󠀳
4. Identifikasi penulis dengan rating rata-rata buku tertinggi: temukan buku dengan rating minimal 50.󠀲󠀡󠀠󠀦󠀥󠀩󠀡󠀢󠀳
5. Hitung jumlah rata-rata teks ulasan di antara pengguna yang memberi rating terhadap lebih dari 50 buku.

# Deskripsi data

books:Berisi data tentang buku:

- book_id — ID buku
- author_id — ID penulis
- title — judul buku
- num_pages — jumlah halaman
- publication_date — tanggal penerbitan
- publisher_id — ID penerbit

authors:Berisi data tentang penulis:

- author_id — ID penulis
- author — nama penulis
- publishers:

Berisi data tentang penerbit:

- publisher_id — ID penerbit
- publisher — nama penerbit
- ratings:

󠀰Berisi data tentang rating pengguna:

- rating_id — ID rating
- book_id — ID buku
- username — nama pengguna yang memberi rating buku
- rating
- reviews:

Berisi data tentang ulasan pelanggan:

- review_id — ID ulasan
- book_id — ID buku
- username — nama pengguna yang mengulas buku
- text — teks ulasan

# Import Library

In [1]:
# impor library yang diperlukan
import pandas as pd
from sqlalchemy import create_engine
from IPython.display import Markdown, display

db_config = {'user': 'practicum_student',         # nama pengguna
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # kata sandi
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # port koneksi
             'db': 'data-analyst-final-project-db'}          # nama database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Eksplorasi Data

In [2]:
def query(query_string):
    return pd.io.sql.read_sql(query_string, con = engine)

In [3]:
# Daftar tabel dari database
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']

# Loop untuk membaca data dan menampilkan informasi lima baris pertama
for table in tables:
    print(f"Table: {table}")
    
    query_string = f"SELECT * FROM {table}"
    data = query(query_string)
    display(data.info())
    display(data.head())
    
    print("\n")

Table: books
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268



Table: authors
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd



Table: publishers
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company



Table: ratings
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2



Table: reviews
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

- books : terdiri dari 1000 baris 
- authors : terdiri dari 636 baris
- publishers :terdiri dari 340 baris
- ratings : terdiri dari 6456 baris 
- reviews : terdiri dari 2793 baris

# Pengolahan Data

In [4]:
#tugas 1 : Hitung jumlah buku yang dirilis setelah tanggal 1 Januari 2000.󠀲󠀡󠀠󠀦󠀥󠀩󠀠󠀩󠀳
query_books_after_2000 = pd.io.sql.read_sql('''
SELECT COUNT (DISTINCT book_id)
FROM books
WHERE publication_date > '2000-01-01'
''', con = engine)

display(query_books_after_2000)

count
0    819

Berdasarkan pengolahan, ditemukan bahwasanya terdapat 819 buku yang dirilis setelah tanggal 1 Januari 2000

In [5]:
#tugas 2 : Hitung jumlah ulasan pengguna dan rating rata-rata untuk setiap buku.󠀲󠀡󠀠󠀦󠀥󠀩󠀡󠀠󠀳
query_mean_reviews = pd.io.sql.read_sql('''
SELECT 
    books.book_id AS book_id,
    books.title AS book_title,
    COALESCE(avg_rating, 0) AS avg_rating,
    COALESCE(number_of_reviews, 0) AS number_of_reviews
FROM 
    books
LEFT JOIN (
    SELECT 
        book_id, 
        AVG(rating) AS avg_rating
    FROM 
        ratings
    GROUP BY 
        book_id
) AS ratings_avg ON books.book_id = ratings_avg.book_id
LEFT JOIN (
    SELECT 
        book_id, 
        COUNT(review_id) AS number_of_reviews
    FROM 
        reviews
    GROUP BY 
        book_id
) AS reviews_count ON books.book_id = reviews_count.book_id
ORDER BY 
    number_of_reviews DESC,
    avg_rating DESC
''', con=engine)
display(query_mean_reviews.head(10))

book_id                                         book_title  avg_rating  \
0      948                            Twilight (Twilight  #1)    3.662500   
1      302  Harry Potter and the Prisoner of Azkaban (Harr...    4.414634   
2      299  Harry Potter and the Chamber of Secrets (Harry...    4.287500   
3      656                                     The Book Thief    4.264151   
4      734                                   The Glass Castle    4.206897   
5      497                          Outlander (Outlander  #1)    4.125000   
6      750                The Hobbit  or There and Back Again    4.125000   
7      695  The Curious Incident of the Dog in the Night-Time    4.081081   
8      779  The Lightning Thief (Percy Jackson and the Oly...    4.080645   
9      963                                Water for Elephants    3.977273   

   number_of_reviews  
0                  7  
1                  6  
2                  6  
3                  6  
4                  6  
5                  6  
6                  6  
7                  6  
8                  6  
9                  6

- Left join digunakan untuk mempertahankan jumlah buku yang direkap pada tabel books
- buku dengan number of reviews tertinggi (7) ialah buku Twilight dengan rata-rata rating 3.6 

In [6]:
#tugas 3 : Identifikasi penerbit yang telah menerbitkan jumlah buku terbanyak dengan jumlah halaman lebih dari 50 (hal ini akan membantumu dalam mengecualikan brosur dan publikasi serupa dari analisismu).󠀲󠀡󠀠󠀦󠀥󠀩󠀡󠀡󠀳
query_page_above50 = pd.io.sql.read_sql('''
SELECT 
    publishers.publisher AS publisher_name,
    COUNT(books.book_id) AS total_books_published
FROM 
    books
JOIN 
    publishers ON books.publisher_id = publishers.publisher_id
WHERE 
    books.num_pages > 50
GROUP BY 
    publishers.publisher_id
ORDER BY 
    total_books_published DESC
LIMIT 10
''', con = engine)

display(query_page_above50)

publisher_name  total_books_published
0             Penguin Books                     42
1                   Vintage                     31
2  Grand Central Publishing                     25
3          Penguin Classics                     24
4          Ballantine Books                     19
5                    Bantam                     19
6                   Berkley                     17
7        St. Martin's Press                     14
8             Berkley Books                     14
9                     Delta                     13

Penguin books merupakan publikasi yang memiliki jumlah buku yang telah diterbitkan terbanyak yakni mencapai 42 buku. Tabel diatas merupakan 5 nama publikasi yang memiliki jumlah terbitan buku tertinggi

In [7]:
#tugas 4: Identifikasi penulis dengan rating rata-rata buku tertinggi: temukan buku dengan rating minimal 50.󠀲󠀡󠀠󠀦󠀥󠀩󠀡󠀢󠀳
query_author_rating = pd.io.sql.read_sql('''
SELECT 
    authors.author,
    AVG(subquery2.avg_rating) AS final_avg
FROM (
    SELECT 
        books.title,
        books.author_id,
        subquery1.avg_rating
    FROM (
        SELECT 
            book_id, 
            COUNT(rating_id) AS rating_cnt,
            AVG(rating) AS avg_rating
        FROM 
            ratings
        GROUP BY 
            book_id
        HAVING 
            COUNT(rating_id) > 50) AS subquery1
        INNER JOIN 
            books on books.book_id = subquery1.book_id) AS subquery2
    INNER JOIN authors ON authors.author_id = subquery2.author_id
GROUP BY
    author
ORDER BY 
    final_avg DESC
LIMIT 10
''', con=engine)

display(query_author_rating)

author  final_avg
0                         J.K. Rowling/Mary GrandPré   4.283844
1                  Markus Zusak/Cao Xuân Việt Khương   4.264151
2                                     J.R.R. Tolkien   4.258446
3                                  Louisa May Alcott   4.192308
4                                       Rick Riordan   4.080645
5                                    William Golding   3.901408
6                                      J.D. Salinger   3.825581
7           Paulo Coelho/Alan R. Clarke/Özdemir İnce   3.789474
8  William Shakespeare/Paul Werstine/Barbara A. M...   3.787879
9                                          Dan Brown   3.754540

tabel berikut menampilkan 10 penulis dengan rata-rata rating buku minimal 50. lebih lanjut dari tabel dapat dilihat bahwa penulis buku dengan nama J.K.Rowling memiliki rata-rata rating buku tertinggi yakni 4.28. Markus Zusak berada pada posisi kedua dengan rata-rata rating 4.26. Hanya selisih kurang lebi 0.02 poin

In [8]:
#tugas 5: Hitung jumlah rata-rata teks ulasan di antara pengguna yang memberi rating terhadap lebih dari 50 buku.
query_text_rating= pd.io.sql.read_sql('''
SELECT 
    ROUND(AVG(subquery.text_reviews_count), 2) as avg_text_review_count
FROM(
    SELECT 
        COUNT(DISTINCT reviews.review_id) AS text_reviews_count
    FROM 
        reviews
    INNER JOIN 
        ratings
    ON 
        reviews.username = ratings.username
    GROUP BY 
        ratings.username
    HAVING 
        COUNT(DISTINCT ratings.rating_id) > 50) as subquery
LIMIT 10
''', con=engine) 

display(query_text_rating)

avg_text_review_count
0                  24.33

berdasarkan hasil pengolahan dapat diketahui bahwasanya rata-rata teks ulasan yang diberikan oleh pembaca adalah 24.33 

# Kesimpulan 

Tahapan: Eksplorasi Data
- data SQL terdiri dari 5 tabel yaitu books, authors, publishers, ratings dan review. 
- rekapan data dikethaui menyimpan 1000 judul buku dengan 636 penulis, 340 instansi publikasi. 
- buku-buku tersebut telah diberi nilai oleh 6456 orang dengan 29793 diantaranya meninggalkan kesan pesan pengalaman membaca buku tersebut

Tahapan: Pengolahan Data 
- terdapat 819 buku yang dipublikasi setelah 1 Januari 2000
- Twilight#1 merupakan buku dengan jumlah review terbanyak yaitu 7 orang dengan rating 3.6
- diantara 340 instansi publikasi, hasil karya yang diterbitkan oleh Penguin books berhasil menjadi instansi dengan jumlah buku dirilis terbanyak (42 buku)
-  J.K. Rowling (4.28) dan Markus Zusak (4.26) merupakan 2 penulis buku yang memiliki rata-rata rating buku tertingi, dimana selisih nilai antar keduanya adalah 0.02 
- rata-rata pembaca meningkalkan komentar terhadap buku berkiras 24 kata